File: java/src/day04/Scratchcards.java
package day04;

import java.io.IOException;
import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

public class Scratchcards {
    public static void main(String[] args) {
        try {
            List<String> cards = getFile("inputs/day04/input.txt");
            System.out.println("Part 1: " + part1(cards));
            System.out.println("Part 2: " + part2(cards));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static int part1(List<String> cards) {
        int result = 0;
        for (String card : cards) {
            int numberOfWinningNumbers = getNumberOfWinningNumbers(card);
            if (numberOfWinningNumbers != 0) {
                if (numberOfWinningNumbers == 1) {
                    result += 1;
                } else {
                    int pointsSystem = numberOfWinningNumbers - 1;
                    result += (int) Math.pow(2, pointsSystem);
                }
            }
        }
        return result;
    }

    private static int getNumberOfWinningNumbers(String card) {
        String parsedCard = card.strip().split(": ")[1];
        String[] winningAndMyNumbers = parsedCard.strip().split("\\|");
        List<Integer> winningNumbers = Arrays.stream(winningAndMyNumbers[0].strip().split("\\s+")).map(Integer::valueOf).toList();
        List<Integer> myNumbers = Arrays.stream(winningAndMyNumbers[1].strip().split("\\s+")).map(Integer::valueOf).toList();
        return winningNumbers.stream().filter(myNumbers::contains).toList().size();
    }

    private static int part2(List<String> cards) {
        Map<Integer, Integer> numOfCards = new HashMap<>();
        int numberOfCards = cards.size();
        for (String card : cards) {
            String[] parsedCardAndNumbers = card.strip().split(": ");
            Integer cardNum = Integer.parseInt(parsedCardAndNumbers[0].strip().split("\\s+")[1]);
            int numberOfWinningNumbers = getNumberOfWinningNumbers(card);
            numOfCards.merge(cardNum, 1, Integer::sum);
            int loopForNumOfCards = numOfCards.get(cardNum);
            for (int i = 1; i <= numberOfWinningNumbers; i++) {
                cardNum += 1;
                if (cardNum > numberOfCards) {
                    break;
                }
                numOfCards.merge(cardNum, loopForNumOfCards, Integer::sum);
            }
        }
        return sumOfCards(numOfCards);
    }

    private static int sumOfCards(Map<Integer, Integer> numOfCards) {
        int result = 0;
        for (Integer value: numOfCards.values()) {
            result += value;
        }
        return result;
    }

    private static List<String> getFile(String file) throws IOException {
        Path currentWorkingDir = FileSystems.getDefault().getPath(System.getProperty("user.dir"));
        Path absoluteFilePath = currentWorkingDir.resolve(file);
        try (Stream<String> input = Files.lines(absoluteFilePath)) {
            return input.toList();
        }
    }
}

File: type-script/day04/scratchcards.ts
import * as fs from 'fs';

const f: string = fs.readFileSync("../inputs/day04/input.txt", { encoding: "utf-8" });
const cards: string[] = f.trim().split("\n").map(x => x.split(": ")[1])


function part1(cards: string[]): number {
    let result: number = 0;
    for (const card of cards) {
        const winningFound: number = numberOfWinningNumbers(card);

        let cardValue: number = 0;
        if (winningFound > 0) {
            cardValue = 2 ** (winningFound - 1);
        }
        result += cardValue;
    }

    return result;
}

function part2(cards: string[]): number {
    let idx: number = 0;
    const numOfCards: number[] = new Array(cards.length).fill(1);
    for (const card of cards) {
        const winningFound: number = numberOfWinningNumbers(card);
        const numCard: number = numOfCards[idx];
        idx++;

        for (let i = 0; i < winningFound; i++) {
            numOfCards[idx + i] += numCard;
        }
    }

    return numOfCards.reduce((a, b) => a + b);
}

function numberOfWinningNumbers(card: string): number {
    const splitCard: string[] = card.split("|");
    const myNumbers: number[] = splitCard[0].trim().split(/\s+/).map(x => Number(x));
    const winningNumbers: number[] = splitCard[1].trim().split(/\s+/).map(x => Number(x));

    return myNumbers.filter(num => winningNumbers.includes(num)).length;
}

console.log("Part 1:", part1(cards));
console.log("Part 2:", part2(cards));


